cpp/tests/io/arrow_io_source_test.cpp

```diff
@@ -87,19 +88,16 @@ TEST_F(ArrowIOTest, S3FileSystem)
     ASSERT_EQ(1, tbl.tbl->num_columns());  // Only single column specified in reader_options
     ASSERT_EQ(244, tbl.tbl->num_rows());   // known number of rows from the S3 file
   }
+
+#ifdef ARROW_S3
```

Comment:
I think we had some issues before with this, and decided it needs to be a runtime check. If you have a precompiled Arrow, you don't get this definition and you don't know how your precompiled Arrow was built, right?

See https://github.com/rapidsai/cudf/issues/11559 for related discussion.

Reply:
(Note that this is using `ARROW_S3` rather than `S3_ENABLED`, so it might be different.)

Reply:
> (Note that this is using `ARROW_S3` rather than `S3_ENABLED`, so it might be different.)

Correct. This is the big difference compared to #11559 ( which steered me towards using dlopen ). `ARROW_S3` is a define provided by arrow itself ( not CMake ) and maps to the exact specification on how Arrow was built. So in this case we won't have any mismatches.

Reply:
Okay good! Approving.

Reply:
Ohhh it's in `<arrow/util/config.h>`. I overlooked that above. That makes lots more sense.